01-architecture.md

```diff
@@ -52,38 +52,55 @@ One key constraint is that actor code size/footprint should be kept as low as po
-      - Supports VMs that rely on internal dispatch (e.g. EVM).
-      - Environment stays agnostic to actor evolution.
-      - Simpler; removes complexity from the environment.
-      - Higher flexibility to accommodate sophisticated method dispatch, e.g. structural pattern matching, logic versioning, low-level proxying, etc.
-    - Cons:
-      - Because the calling convention is VM-specific, in cross-contract calls, the caller needs to know the VM type of the callee so it can pack the call adequately.
-
-We lean towards (2), and this places some constraints on usability:
-
-- SDK should provide utilities to facilitate internal dispatch based on method number _and_ VM's calling convention.
-- SDK should provide functions to pack message calls to foreign actors.
+## Actor public interface
+
+We considered two main approaches. Hybrids of these two approaches were also evaluated.
+
+**Approach 1 is external method dispatch.** The actor exports a table of callable methods. The sandbox matches the message to an exported method, either through the `MethodNum` message field, or by interpreting the input parameters string against a calling convention. This is largely how the current VM works today.
+
+**Approach 2 is for actors to expose a single entrypoint and rely on internal method dispatch.** This awards maximal degrees of freedom and sophistication to support techniques such as structural pattern matching, efficient pass-through proxying, interception, and more.
+
+Because internal dispatch will rapidly become boilerplate, the FVM SDK should offer utilities.
+
+Architecturally speaking, Approach 2 is more aligned with the actor model as implemented in the industry (Erlang, Akka, etc.), in which actors interact through single inboxes where messages are deposited.
+
+Approach 2 is simpler, and places no constraints on the sandbox. It is easier to reason about and reduces overall complexity. It's also more performant because the sandbox has no need to analyse the WASM exports table. Evolving actors over time (such as introducing interface changes) is also easier to reason about.
+
+Finally, this approach is readily compatible with VMs that rely on internal dispatch (e.g. EVM), but also stretches to accommodates for VMs that perform outer dispatch.
+
+Approach 2 would render the `MethodNum` field on messages obsolete, as external dispatch is no longer necessary, and all call information would be contained in the input parameters.
+
+We _could_ technically preserve the `MethodNum`, and have actors perform internal dispatch based on that _and_ input parameters. But we believe that's unnecessary cruft, and it breaks the actor-orientation paradigm by carrying over a procedural construct.
+
+Not everything is rosy, though. The main risk with Approach 2 is that the calling convention not explicited, which may lead to proliferation of calling conventions across user-deployed actors. In turn this may result in cognitive overhead and interoperability issues.
```

Comment:
(Minor typo: `explicited` -> `explicit`)

> proliferation of calling conventions across user-deployed actors

Is it possible to nudge in the right direction by "making the right thing the easy thing" in the SDK? It doesn't solve it right at the root, but could help for the majority of cases. Elsewhere there's been talk about an ERC-like collection of standard interfaces, which could help introduce standard calling conventions in a bottom-up way.

Reply:
Given the IDL, I think we can just make this a non-issue. That is, the actor's interface description would describe the method call "template" in a way that's completely machine readable.

Reply:
👍 for "making the right thing easy" via nice SDKs

Reply:
@Stebalien correct. With a well-writen IDL and good tooling support, it should be possible to pack the call into the right template.

Reply:
Issue to track the IDL: https://github.com/filecoin-project/fvm-project/issues/29